calico/felix/endpoint.py

```diff
@@ -81,6 +102,86 @@ def _create(self, combined_id):
                              self.fip_manager,
                              self.status_reporter)
 
+    @actor_message()
+    def on_tier_data_update(self, tier, data):
+        """
+        Message received when the metadata for a policy tier is updated
+        in etcd.
+
+        :param str tier: The name of the tier.
+        :param dict|NoneType data: The dict or None, for a deletion.
+        """
+        _log.debug("Data for policy tier %s updated to %s", tier, data)
+
+        # Currently, the only data we care about is the order.
+        order = None if data is None else data["order"]
+        if self.tier_orders.get(tier) == order:
+            _log.debug("No change, ignoring")
+            return
+
+        if order is not None:
+            self.tier_orders[tier] = order
+        else:
+            del self.tier_orders[tier]
+
+        new_tier_sequence = sorted(self.tier_orders.iterkeys(),
+                                   key=lambda k: (self.tier_orders[k], k))
+        if self.tier_sequence != new_tier_sequence:
+            _log.info("Sequence of profile tiers changed, refreshing all "
+                      "endpoints")
+            self.tier_sequence = new_tier_sequence
+            self.endpoints_with_dirty_policy.update(
+                self.endpoints_by_id.keys()
+            )
+            self._update_dirty_policy()
+
+    @actor_message()
+    def on_prof_labels_set(self, profile_id, labels):
+        _log.debug("Profile labels updated for %s: %s", profile_id, labels)
+        self._label_inherit_idx.on_parent_labels_update(profile_id, labels)
+        self._update_dirty_policy()
+
+    @actor_message()
+    def on_policy_selector_update(self, policy_id, selector_or_none,
+                                  order_or_none):
+        _log.debug("Policy %s selector updated to %s (%s)", policy_id,
+                   selector_or_none, order_or_none)
+        if selector_or_none != self.selectors_by_pol_id.get(policy_id):
+            if selector_or_none is not None:
+                self.selectors_by_pol_id[policy_id] = selector_or_none
+            else:
+                del self.selectors_by_pol_id[policy_id]
+            # Defer to the label index, which will call us back via
+            # on_policy_match_started and on_policy_match_stopped.
+            self.policy_index.on_expression_update(policy_id,
+                                                   selector_or_none)
```

Comment:
I believe you're relying here on synchronous callbacks to on_policy_match_started and on_policy_match_stopped, such that those callbacks happen before the _update_dirty_policy call below.  (a) Is it worth commenting that?  (b) Are there any code paths that can lead to calls recursing many times (or indefinitely)?